ENT-4601 Public API to run external operations from a flow

lankydan opened the pull request:
Deprecate `FlowAsyncOperation` and reimplement public versions `FlowExternalOperation` and `FlowExternalAsyncOperation`.

`await` added to `FlowLogic` to allow easy calling from both Java and Kotlin. There are two overrides of `await` (one for `FlowExternalOperation` and `FlowExternalAsyncOperation`).

Implementations of `FlowExternalOperation` return a result (written as _blocking_ code) from their `execute` function. This operation will then be executed using a thread provided by the `externalOperationExecutor`.

Implementations of `FlowExternalAsyncOperation` return a future from their `execute` function. This operation must be executed on a newly spawned thread or one provided by a thread pool. It is up to developers to handle threading in this scenario.

The default thread pool (`externalOperationExecutor`) can be configured through the `flowExternalOperationThreadPoolSize` node config.

The current implementation leaves `FlowAsyncOperation` alone, meaning that any developers that have used it (even though it is internal) won't need to change their apps. If this was not concern I would delete it completely and replumb the state machine code. Instead, it has been marked with `@DoNotImplement` and `executeAsync` is annotated with `@Deprecated`

lankydan committed:
ENT-4601 First cut of improving flow async api

- Create `FlowBackgroundProcess` to replace `FlowAsyncOperation`
- Add `doOnComplete` to `CordaFutureImpl` to improve tests (should be
  useful for users as well)
- Remove redundant logging in `TransitionExecutorImpl`

lankydan committed:
ENT-4601 Rename async functions to `await`

- Naming convention of `await` and `awaitFuture`
- Pass in `ServiceHub` to functions to resolve a `NullPointerException` that is thrown
  when a `await` function directly accesses a flow's `serviceHub`. The `serviceHub` is
  injected into an anonymous object that hold's the reference to the `serviceHub`. The
  `serviceHub` is then a parameter to `operation` function inside of `await` and
  `awaitFuture`. Using this `serviceHub` circumvents this issue.
- Separate tests out

lankydan committed:
ENT-4601 Remove unwanted `await` implementations

lankydan committed:
ENT-4601 Provide configurable thread pool for background processes

- Add an `ExecutorService` whose size can be configured by node config
- The thread pool size is controlled by the `flowBackgroundProcessThreadPoolSize`
  config option
- The default thread pool size is equal to the number of available processors
- Add a reference to the background process `ExecutorService` to
  `ServiceHubCoreInternal`. This allows it to be referenced by the core code,
  while the thread pool can be instantiated by code living in node.

lankydan committed:
ENT-4601 Fix detekt errors

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/FlowAsyncOperation.kt:
```diff
@@ -25,6 +27,10 @@ interface FlowAsyncOperation<R : Any> {
 
 // DOCSTART executeAsync
```
This implies it appears in the docs somewhere.... interesting.  Can you track it down?  We should replace any docs with the new API.

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FlowBackgroundProcess.kt:
```diff
@@ -0,0 +1,73 @@
+// create extension functions to be able to call the same code from within a service??
+
+// does this need to be @CordaSerializable?
+interface FlowBackgroundProcess<R : Any> : FlowAsyncOperation<R>
```
Let's call this `FlowExternalOperation`.  I'm keen to emphasise this is "off flow".  i.e. you are not "running in the flow".

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FlowBackgroundProcess.kt:
```diff
@@ -0,0 +1,73 @@
+// does this need to be @CordaSerializable?
+interface FlowBackgroundProcess<R : Any> : FlowAsyncOperation<R>
+
+private abstract class FlowBackgroundProcessImpl<R : Any>(internal val serviceHub: ServiceHub) : FlowBackgroundProcess<R>
```
I don't think we should provide/encourage/allow `ServiceHub` access outside of the flow.  I'm tempted to explicitly ban it (somehow).  Certainly explicitly document they should not capture it in the operation.  It will only increase the chances of a database error happening in an environment where there is no flow hospital etc.  If they want to access `ServiceHub`, they should do it in the flow.  This is supposed to be for doing stuff that is external.

Additionally, the number of database connections is derived from the number of threads we expect to access the database.  So if we allow database access, we need to be increasing the connection pool auto sizing.

rick-r3 replied:
`AbstractFlowExternalOperation`, although maybe existence unnecessary without `ServiceHub`.

lankydan replied:
We need to provide a way for developers to create their own thread pools for creating their own futures. The only way for them to create a thread pool is through a service. Therefore, they will need the service hub to be able to access them. I think that is correct.

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/FlowAsyncOperation.kt:
```diff
@@ -10,6 +11,7 @@ import net.corda.core.serialization.CordaSerializable
  * Interface for arbitrary operations that can be invoked in a flow asynchronously - the flow will suspend until the
  * operation completes. Operation parameters are expected to be injected via constructor.
  */
+@DoNotImplement
```
Not strictly required due to the `internal` package.  All a bit subtle, I know.

lankydan replied:
Wanted to fully indicate to anyone using it (even though it is internal) that they shouldn't be using it. I could deprecate it but it looks weird since all of the new code is just using it behind the scenes.

lankydan committed:
ENT-4601 Add tests around accessing database from background process

lankydan committed:
ENT-4601 Fix tests by adding mock thread pool size

dimosr commented:
When I was looking at this piece of work several weeks ago, I had the following thought: should we enhance this API now that it will become public with an additional method (e.g. `cleanup(deduplicationId)`) that's invoked by Corda? I think use-cases that need exactly-once guarantees and have to keep track of dedup IDs would benefit from a clear indication on when it's safe to remove them.

lankydan commented:
@dimosr Possibly. I talked to @mnesbit about adding inbuilt support inside of corda for handling deduplication ids (1 version where corda handles, 1 where it does not) but we decided not to add this in (at the moment at least). A `cleanup` phase might be useful but I would need to see how that would affect my new api functions, would make them uglier but they should still work fine.

dimosr commented:
Hmm, my understanding is that deduplication needs to be done on the external system where side-effects take place in order to get exactly-once guarantees. I don't see how Corda could handle this.

lankydan committed:
ENT-4601 Test correction

lankydan committed:
ENT-4601 Test error inside of `await` + a database transaction

mnesbit commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -336,6 +337,8 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
     protected abstract val rxIoScheduler: Scheduler
 
+    val backgroundProcessExecutor = createBackgroundProcessExecutor(configuration.flowBackgroundProcessThreadPoolSize)
+
```
Not sure what DB context this stuff will operate in. If people call ServiceHub then I agree with @rick-r3 that it will blow up.

lankydan replied:
Yes, I have been working off the assumption that you would need to write something like the following: 

```kotlin
// inside a corda service
fun saveToDatabaseWithDatabaseTransaction(entity: CustomTableEntity): Unit = 
  services.database.transaction {
    session.save(entity)
  }
```

That opens and commits a local database transaction or uses the flows database transaction if not called from within an `await` block.

lankydan committed:
ENT-4601 Fix test that I actually broke when trying to correct it...

lankydan committed:
ENT-4601 Move code into `FlowLogic` and improve Java API

Move the `await` functions into `FlowLogic` instead of relying on kotlin
extension functions. This means that they can be called from Java is a nicer
looking way.

Although the original `await` functions would compile in Java, they would
fail at runtime due to issues with Kyro and Quasar (lambda's need to be
specified as serializable due to being in the suspended stack). To get around
this a new interface `SerializableBiFunction` was created that was a function
and also extended `Serializable`. Overrides of `await` and `awaitFuture` were
added that used this new interface.
  - Kotlin now uses the original version (taking a function)
  - Java uses the new version (taking in `SerializableBiFunction`)

Tests were added to check the Java API.

lankydan committed:
ENT-4601 `SerializableFunctions.kt` to `SerializableBiFunction.kt` to appease detekt

rick-r3 commented:
I think @dimosr might be correct here.  There was something we never added to the original API and I couldn't remember what it was.  I think it might have been to hook it into the `DeduplicationHandler` logic.  Basically, after we successfully take the next checkpoint, we fire the de-duplication handler (`afterDatabaseTransaction`) to notify the "background process" that there is no chance of it seeing the same de-duplication ID again, and can thus clean up any state associated with de-duplication.  That might not be quite right, but that's the sort of concept.

I think it would mean adding a new method along side `execute` on the `FlowAsyncOperation` (but perhaps your new sub-interface so existing users not affected) that looked a bit like:

`fun cleanup(deduplicationId: String)`

We can chat about it Monday, but I think we shouldn't creep the scope here and put that as a potential future enhancement (since it looks like an additional method).  Can you capture this in a JIRA so we don't forget what we meant this time please @lankydan ?

lankydan committed:
ENT-4601 Do not use lambda `await` implementations

To make it harder to use code from `ServiceHub`, the `await` implementations
that used lambdas have been removed. The theory behind this is that if a
developer wants to access the `ServiceHub` they still can, but will need
to explicitly pass it into their own implementation. Developers __can__
still use the `ServiceHub` but need to go through the effort to do so.

`FlowBackgroundProcess` renamed to `FlowExternalFuture`

`FlowExternalResult` interface made which developers implement with a normal
function which will be ran on the default thread pool. `FlowExternalResult`
is converted into a `FlowExternalFuture` under the hood.

lankydan committed:
ENT-4601 Return `CompletableFuture` from `FlowExternalFuture`

Replace the `CordaFuture` returned from `FlowExternalFuture.execute` with a
`CompletableFuture`. This is to allow developers to use the functions that
Java provides out of the box for handling the creation and completion of
futures.

Part of the decision to do this, is the fact that all of the `CordaFuture`
functions are part of the internal API.

Using `CompletableFuture` is easier from a users perspective as they would
likely create one anyway, and then wrap/convert it to a `CordaFuture`.

lankydan committed:
ENT-4601 Rename Java test

lankydan committed:
ENT-4601 Rename `backgroundProcessExecutor` to `externalOperationExecutor`

lankydan commented:
Raised https://r3-cev.atlassian.net/browse/ENT-4842 to address deduplication id cleanup

lankydan commented:
run smoke tests

rick-r3 commented on core/src/main/kotlin/net/corda/core/flows/FlowExternalFuture.kt:
```diff
@@ -0,0 +1,64 @@
+ * have been processed. It is possible to interact with the database from inside a [FlowExternalFuture] but, for most operations, is not
+ * currently supported.
+ */
+interface FlowExternalFuture<R : Any?> {
```
I'm not sure about the naming of `FlowExternalFuture` and `FlowExternalResult `.  The suffixes are the problem.  The future is not a `Future` and the result is not a result.

`FlowExternalCallable` (but it doesn't really match the `Callable` interface) or `FlowExternalOperation` instead of `FlowExternalFuture`.

I suppose the future one could be `FlowExternalAsyncOperation`.

lankydan replied:
I couldn't really come up with any other names for these interfaces. The best I could think of were `FlowExternalOperation` (future version) and `BlockingFlowExternalOperation` (_blocking_ version).

lankydan replied:
Went through some ideas with @JamesHR3 we managed to come up with `FlowExternalSchedulableOperation` (future version) and `FlowExternalOperation` (blocking version). It is a bit of a mouthful but we think it gives a decent explanation on what it does.

rick-r3 commented:
Other than the naming, I think it looks fine now.

lankydan committed:
ENT-4601 Write KDocs for new API

lankydan committed:
Remove `SerializableBiFunction`

lankydan committed:
ENT-4601 Updated KDoc on `FlowExternalFuture`

lankydan committed:
ENT-4601 Rename to `FlowExternalOperation` and `FlowExternalAsyncOperation`

rick-r3 reviewed (dismissed):
But needs rebase...

lankydan committed:
ENT-4601 Documentation on using calling external systems from flows

`api-flows` added the following sections
  - `Calling external systems inside of flows`
    - FlowExternalOperation
    - FlowExternalAsyncOperation
    - Handling deduplication in external operations
    - Creating CompletableFutures from Guava's ListenableFutures

`corda-configuration-file` updated to mention `flowExternalOperationThreadPoolSize`

lankydan committed:
ENT-4601 Update docs to emphasise deduplication handling on the external system's side

lankydan committed:
ENT-4601 Remove documentation of internal `FlowAsyncOperation`

lankydan committed:
ENT-4601 Tidy up tests

lankydan committed:
ENT-4601 Fix detekt

lankydan commented:
When squashing this commit, please use the description of the PR only (too many messy commits).

lankydan committed:
ENT-4601 Remove calls to deprecated `executeAsync` function

Also tidy up checkpoint dumper code to allow serialization
of the new interfaces added for external operations.

lankydan committed:
ENT-4601 Update API

lankydan committed:
ENT-4601 Tell people not to use `GetHospitalCountersFlow`

vkolomeyko commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -731,6 +734,17 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
+        }
+        return Executors.newFixedThreadPool(
+            numberOfThreads,
+            ThreadFactoryBuilder().setNameFormat("flow-external-operation-thread").build()
```
I have fallen into this trap as well, if `ThreadFactoryBuilder` is not explicitly specifying `daemon` setting then:
"is a daemon thread if and only if the creating thread is a daemon."
That said threads at this executor may not be daemons which might be undesired.
